### PR TITLE
Correctly handle empty strings in params of RemoteExecutorTestBase

### LINKED ISSUE #21181: RemoteExecutorTestBase throws TargetParameterCountException passing an empty string as an argument
Example:
```cs
RemoteInvoke(param => Assert.Empty(param), string.Empty).Dispose();
```

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #21181

I tested this by modifying an existing test in System.Runtime to use the empty string alongside other data.

As part of this, I noticed that one of the tests that asserts a FileNotFoundException actually originated from Assembly.LoadFrom, rather than the actual method that we were testing. We already have coverage for Assembly.LoadFrom where the path doesn't exist, so we can safely delete that test.

### CODE REVIEW

File: src/System.Runtime/tests/System/TypeTests.cs [hunk 1]
```diff
@@ -340,38 +340,19 @@ public static void GetTypeByName()
...
-               {
-                   string test1 = testtype;
-                   //Loading from the wrong path
-                   Assert.Throws<System.IO.FileNotFoundException>(() =>
-                   Type.GetType(test1,
-                     (aName) => aName.Name == "Foo" ?
-                         Assembly.LoadFrom(@".\NoSuchTestLoadAssembly.dll") : null,
-                     typeloader,
-                     true
-                  ));
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/TypeTests.cs]: Why did you delete this test?

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs [hunk 2]
```diff
@@ -49,13 +50,28 @@ private static RemoteInvokeHandle RemoteInvoke(MethodInfo method, string[] args,
...
+            testConsoleAppArgs.Append(' ');
+            testConsoleAppArgs.Append(method.Name);
+
+            // We need to surround each argument with quotes.
+            foreach (string arg in args)
+            {
+                testConsoleAppArgs.Append(" \"");
+                testConsoleAppArgs.Append(arg);
+                testConsoleAppArgs.Append('"');
+            }
```
Comment (stephentoub) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: Are there uses of RemoteInvoke manually wrapping arguments in quotes today?  If so, can you remove those?

File: src/System.Runtime/tests/System/TypeTests.cs (hunk 1 above)
Comment (danmosemsft) [inline src/System.Runtime/tests/System/TypeTests.cs]: see above

Comment (stephentoub) [inline src/System.Runtime/tests/System/TypeTests.cs]: Ah, fair enough.

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs [hunk 3]
```diff
@@ -49,13 +50,28 @@ private static RemoteInvokeHandle RemoteInvoke(MethodInfo method, string[] args,
             }
 
             // If we need the host (if it exists), use it, otherwise target the console app directly.
-            string testConsoleAppArgs = "\"" + a.FullName + "\" " + t.FullName + " " + method.Name + " " + string.Join(" ", args);
+            var testConsoleAppArgs = new StringBuilder();
+            testConsoleAppArgs.Append(ExtraParameter);
```
Comment (danmosemsft) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: This might be cleaner if you created an "AppendQuoted" method .. it could also escape existing quotes.

File: src/Common/tests/System/Diagnostics/RemoteExecutorConsoleApp/RemoteExecutorConsoleApp.cs [hunk 4]
```diff
@@ -33,6 +33,16 @@ private static int Main(string[] args)
...
 
+            // Unwrap any parameters wrapped in quotation marks.
+            for (int i = 0; i < additionalArgs.Length; i++)
+            {
+                string arg = additionalArgs[i];
+                if (arg.Length >= 2 && arg[0] == '"' && arg[arg.Length -1] == '"')
+                {
+                    additionalArgs[i] = arg.Substring(1, arg.Length - 2);
+                }
+            }
```
Comment (stephentoub) [inline src/Common/tests/System/Diagnostics/RemoteExecutorConsoleApp/RemoteExecutorConsoleApp.cs]: Is this actually necessary?  I'm surprised they're still wrapped in quotes at this point... I'd have expected those quotes to be removed by the shell.  Is that not happening?

**Comment by hughbe** [comment]
It seems that Unix isn't working for empty strings with.
This isn't something I would expect: https://ci.dot.net/job/dotnet_corefx/job/master/job/osx10.12_debug_prtest/4888/

Is RemoteExecutorTestBase different for Windows and Unix?

```
...

04:38:20   Exception from RemoteExecutorConsoleApp(System.Globalization.Tests, Version=4.1.1.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb, System.Globalization.Tests.NumberFormatInfoCurrentInfo+<>c, <CurrentInfo_CustomCulture>b__1_0):
04:38:20   Assembly: System.Globalization.Tests, Version=4.1.1.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb
04:38:20   Type: System.Globalization.Tests.NumberFormatInfoCurrentInfo+<>c
04:38:20   Method: Int32 <CurrentInfo_CustomCulture>b__1_0(System.String)
04:38:20   Using as the test runtime folder.
04:38:20   ls: /cores: No such file or directory
04:38:20   Running tests... Start time: 04:38:20
04:38:20   Commands:
04:38:20   python DumplingHelper.py install_dumpling
04:38:20   __TIMESTAMP=`python DumplingHelper.py get_timestamp`
04:38:20   chmod +x /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/bin/testhost/netcoreapp-OSX-Debug-x64//dotnet
04:38:20   /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/bin/testhost/netcoreapp-OSX-Debug-x64//dotnet xunit.console.netcore.exe System.Diagnostics.DiagnosticSource.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapptests -notrait category=nonosxtests  -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing
04:38:20   python DumplingHelper.py collect_dump $\ `pwd`  System.Diagnostics.DiagnosticSource.Tests /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/bin/runtime/netcoreapp-OSX-Debug-x64/,/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/bin/AnyOS.AnyCPU.Debug/System.Diagnostics.DiagnosticSource.Tests/netstandard/
04:38:20   /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/bin/AnyOS.AnyCPU.Debug/System.Diagnostics.DiagnosticSource.Tests/netstandard /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/src/System.Diagnostics.DiagnosticSource/tests
04:38:20   Finished:    System.Data.Common.Tests
04:38:20   
04:38:20   === TEST EXECUTION SUMMARY ===
04:38:20   Exception: System.Reflection.TargetParameterCountException: Parameter count mismatch.
04:38:20      at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
04:38:20      System.Data.Common.Tests  Total: 1701, Errors: 0, Failed: 0, Skipped: 0, Time: 11.184s
04:38:20      at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
04:38:20      at RemoteExecutorConsoleApp.Program.Main(String[] args) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx10.12_debug_prtest/src/Common/tests/System/Diagnostics/RemoteExecutorConsoleApp/RemoteExecutorConsoleApp.cs:line 55
04:38:20   
04:38:20   Unhandled Exception: System.Reflection.TargetParameterCountException: Parameter count mismatch.
04:38:20      at RemoteExecutorConsoleApp.Program.Main(String[] args) in /Users/dotnet-
...
```

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs (hunk 3 above)
Comment (AtsushiKan) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: RemoteInvoke() should implement the canonical tokenizing rules - that way, tests can just pass the args they want and not worry about them coming out wrong on the other side. Perhaps there could be a "verbatim" options for tests that specifically want to pass weird command lines to test the tokenizing. We don't need to worry about that here, though.

Here's an implementation I've been using for my private projects for years:

https://github.com/AtsushiKan/test/blob/master/Paste.cs

This will also be interesting to see if Diagnostic.Process tokenizes differently on Unix systems.

Comment (hughbe) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: I added an option earlier that controls whether we quote each argument (on by default)

> This will also be interesting to see if Diagnostic.Process tokenizes differently on Unix systems.

It does look like the Unix builds are failing with TargetParameterCountException. I don't really know why - do you have any ideas?

Comment (AtsushiKan) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: >I added an option earlier that controls whether we quote each argument (on by default)

Rather than making it an option, I would just add a separate overload (**RemoteInvokeRaw(method, String rawCommandLine, options)**) for the tiny subset of tests that actually care about passing untokenized command lines. I think that would be the Environment.CommandLine tests and System.Diagnostic.Process tests that focus on testing tokenization and nothing else.

For everyone else, it should "just work" - they pass in an array of arguments containing arbitrary characters and RemoteInvoke owns tokenizing it in a roundtrippable way.

Comment (AtsushiKan) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: >It does look like the Unix builds are failing with TargetParameterCountException. I don't really know why - do you have any ideas?

Yes - it means you unconvered a bug in the Unix version of System.Diagnostic.Process. Here's the the suspect code:

https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L450

and here:

https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L518

you can see that this tokenizer is trying to implement the canonical behavior but gets it wrong for empty strings.

I think it's best to treat this as a blocking product issue and resolve that before attempting to go forward with this.

Comment (AtsushiKan) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: >I would just add a separate overload (RemoteInvokeRaw(method, String rawCommandLine, options)) 

It looks like that overload already exists (https://github.com/dotnet/corefx/blob/master/src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.cs#L120). Though it has some incorrect doc comments.

**Comment by AtsushiKan** [comment]
I've opened a new issue (https://github.com/dotnet/corefx/issues/21267) regarding all these command line tokenization inconsistencies.

**Comment by tarekgh** [comment]
cc @krwq

**Comment by danmosemsft** [comment]
IIUC the UAP remote executor does not use regular command line parsing -- it is a method call to a "message passing" system that goes to the other process.

**Comment by hughbe** [comment]
I'll safely assume this pr is blocked until that issue is fixed

**Comment by tarekgh** [comment]
CC @Priya91 for the System.Diagnostics.Process issue

**Comment by Priya91** [comment]
Process test change - LGTM

File: src/System.Text.RegularExpressions/tests/Regex.Groups.Tests.cs [hunk 5]
```diff
@@ -631,15 +631,15 @@ public void Groups(string pattern, string input, RegexOptions options, CultureIn
...
 
             RemoteInvoke((innerPatternEnc, innerInputEnc, innerOptionsEnc, innerCultureInfoEnc, innerExpectedGroupsEnc) =>
             {
                 string innerPattern = Encoding.UTF8.GetString(Convert.FromBase64String(innerPatternEnc));
                 string innerInput = Encoding.UTF8.GetString(Convert.FromBase64String(innerInputEnc));
                 RegexOptions innerOptions = (RegexOptions)int.Parse(innerOptionsEnc);
                 CultureInfo innerCultureInfo = innerCultureInfoEnc != EmptyPlaceholder ? new CultureInfo(innerCultureInfoEnc) : null;
-                string[] innerExpectedGroups = innerExpectedGroupsEnc != EmptyPlaceholder ? Encoding.UTF8.GetString(Convert.FromBase64String(innerExpectedGroupsEnc.Trim('"'))).Split(Seperator).Select(s => s == EmptyPlaceholder ? string.Empty : s).ToArray() : new string[] { };
+                string[] innerExpectedGroups = innerExpectedGroupsEnc != EmptyPlaceholder ? Encoding.UTF8.GetString(Convert.FromBase64String(innerExpectedGroupsEnc)).Split(Seperator).Select(s => s == EmptyPlaceholder ? string.Empty : s).ToArray() : new string[] { };
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/tests/Regex.Groups.Tests.cs]: nit, typo "Separator"

File: src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs [hunk 6]
```diff
@@ -287,7 +287,7 @@ public void CurrentDirectory()
                 Assert.NotEmpty(Directory.EnumerateFiles(Directory.GetCurrentDirectory(), "*", SearchOption.TopDirectoryOnly));
 
                 return SuccessExitCode;
-            }, $"\"{testDir}\"").Dispose();
+            }, testDir).Dispose();
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs]: these already fixed by 7d5d44a6a905624b066a768c68e413b6d40be9e0

**Comment by danmosemsft** [comment]
> I'll safely assume this pr is blocked until that issue is fixed

@hughbe just to be clear are you looking at that or waiting on someone ? :)

**Comment by hughbe** [comment]
WAiting if you don't mind - sorry!

**Comment by Priya91** [comment]
> WAiting if you don't mind - sorry!

I'm looking at this.

**Comment by danmosemsft** [comment]
Fixes https://github.com/dotnet/corefx/issues/21276

**Comment by AtsushiKan** [comment]
@hughbe - when you resume this, can you put the Paste routine under Common\src so that we can use the same source file to fix up Environment.CommandLine as well? I've upgraded my sample to include argv0 processing. The RemoteExecutor doesn't need that (so it can pass "false" for that extra argument) but Environment.CommandLine will.

**Comment by hughbe** [comment]
Test NETFX x86 Release Build

**Comment by danmosemsft** [comment]
@hughbe not sure why but it's failing on desktop as you see: `  Exception: System.ArgumentException: String cannot have zero length.
     at System.Reflection.AssemblyName..ctor(String assemblyName)
     at RemoteExecutorConsoleApp.Program.Main(String[] args) in D:\j\workspace\windows-TGrou---2a8f9c29\src\Common\tests\System\Diagnostics\RemoteExecutorConsoleApp\RemoteExecutorConsoleApp.cs:line 46`

File: src/Common/src/System/PasteArguments.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace System
+{
+    public static class PasteArguments
```
Comment (AtsushiKan) [inline src/Common/src/System/PasteArguments.cs]: Should be changed to "interna"

File: src/Common/src/System/PasteArguments.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
```
Comment (AtsushiKan) [inline src/Common/src/System/PasteArguments.cs]: You've gotten rid of the Linq dependency in the code - the namespace import shouldgo too or it won't go into S.R.Extensions.

File: src/Common/src/System/PasteArguments.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+namespace System
+{
+    public static class PasteArguments
+    {
+         /// <summary>
+        /// Repastes a set of arguments into a linear string that parses back into the originals under pre- or post-2008 VC parsing rules.
+        /// The rules for parsing the executable name (argv[0]) are special, so you must indicate whether the first argument actually is argv[0].
+        /// </summary>
+        public static string Paste(this IEnumerable<string> arguments, bool pasteFirstArgumentUsingArgV0Rules)
```
Comment (AtsushiKan) [inline src/Common/src/System/PasteArguments.cs]: We can make this a regular (non-extension) method - you're not using it that way anyway.

File: src/Common/src/System/PasteArguments.cs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                if (char.IsWhiteSpace(c) || c == Quote)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private const char Nul = (char)0;
```
Comment (AtsushiKan) [inline src/Common/src/System/PasteArguments.cs]: The "Nul" was left over from a previous 0-terminated version - this isn't used anymore.

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.uap.cs [hunk 11]
```diff
@@ -23,7 +23,8 @@ public abstract partial class RemoteExecutorTestBase : FileCleanupTestBase
         /// <param name="args">The arguments to pass to the method.</param>
         /// <param name="start">true if this function should Start the Process; false if that responsibility is left up to the caller.</param>
         /// <param name="psi">The ProcessStartInfo to use, or null for a default.</param>
-        private static RemoteInvokeHandle RemoteInvoke(MethodInfo method, string[] args, RemoteInvokeOptions options)
+        /// <param name="pastArguments">Unused in UAP.</param>
```
Comment (AtsushiKan) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.uap.cs]: Nit: Typo - "pastArguments"

**Comment by hughbe** [comment]
Addressed all your PR feedback @AtsushiKan :)

**Comment by hughbe** [comment]
Test Linux x64 Release Build

**Comment by danmosemsft** [comment]
@dotnet-bot Test Linux x64 Release Build please

**Comment by hughbe** [comment]
Test Linux x64 Release Build
